mod.go

```diff
@@ -0,0 +1,669 @@
+	cacheMask = cacheSets - 1
+)
+
+type cacheSet struct {
+	hit	uint64
+	miss	uint64
+
+	rw	sync.RWMutex
+	mod	[cacheWays]Int
+	inv	[cacheWays][5]uint64
+}
+
+var cache [cacheSets]cacheSet
+
+func cacheStats() (hit, miss uint64) {
+	for s:=0; s<cacheSets; s++ {
+		hit  += cache[s].hit
+		miss += cache[s].miss
+	}
+	return
+}
+
+// Some utility functions
+
+func add256(x, y Int, c uint64) (Int, uint64) {
+	var z Int
+	z[0], c = bits.Add64(x[0], y[0], c)
+	z[1], c = bits.Add64(x[1], y[1], c)
+	z[2], c = bits.Add64(x[2], y[2], c)
+	z[3], c = bits.Add64(x[3], y[3], c)
+	return z, c
+}
+
+func sub256(x, y Int, b uint64) (Int, uint64) {
+	var z Int
+	z[0], b = bits.Sub64(x[0], y[0], b)
+	z[1], b = bits.Sub64(x[1], y[1], b)
+	z[2], b = bits.Sub64(x[2], y[2], b)
+	z[3], b = bits.Sub64(x[3], y[3], b)
+	return z, b
+}
+
+func leadingZeros(x Int) (z int) {
+	z  = bits.LeadingZeros64(x[3]); if z <  64 { return }
+	z += bits.LeadingZeros64(x[2]); if z < 128 { return }
+	z += bits.LeadingZeros64(x[1]); if z < 192 { return }
+	z += bits.LeadingZeros64(x[0]); return
+}
+
+func onesCount(x Int) (z int) {
+	z =	bits.OnesCount64(x[0]) +
+		bits.OnesCount64(x[1]) +
+		bits.OnesCount64(x[2]) +
+		bits.OnesCount64(x[3])
+	return
+}
+
+func shiftLeft(x Int, s int) (z Int) {
+	if s < 0 {
+		z = shiftright(x, uint(-s))
+	} else {
+		z = shiftleft(x, uint(s))
+	}
+	return
+}
+
+//func shiftRight(x Int, s int) (z Int) {
+//	if s < 0 {
+//		z = shiftleft(x, uint(-s))
+//	} else {
+//		z = shiftright(x, uint(s))
+//	}
+//	return
+//}
+
+func shiftleft(x Int, s uint) (z Int) {
+	w := s/64	// word shift
+	l := s%64	// left shift
+	r := 64-l	// right shift
+
+	var t Int
+
+	// Bit shift
+
+	t[0] = (x[0] << l)
+	t[1] = (x[1] << l) | (x[0] >> r)
+	t[2] = (x[2] << l) | (x[1] >> r)
+	t[3] = (x[3] << l) | (x[2] >> r)
+
+	// Word shift
+
+	i := 0
+	j := w
+
+	for ; j<4; {
+		z[j] = t[i]
+		i++
+		j++
+	}
+
+	return
+}
+
+func shiftright(x Int, s uint) (z Int) {
+	w := s/64	// word shift
+	r := s%64	// right shift
+	l := 64-r	// left shift
+
+	var t Int
+
+	// Bit shift
+
+	t[0] = (x[0] >> r) | (x[1] << l)
+	t[1] = (x[1] >> r) | (x[2] << l)
+	t[2] = (x[2] >> r) | (x[3] << l)
+	t[3] = (x[3] >> r)
+
+	// Word shift
+
+	i := 0
+	j := w
+
+	for ; j<4; {
+		z[i] = t[j]
+		i++
+		j++
+	}
+
+	return
+}
+
+// shiftright320 shifts a 320-bit value 0-63 bits right
+// z = x >> (s % 64)
+
+func shiftright320(x [5]uint64, s uint) (z [5]uint64) {
+	r := s%64	// right shift
+	l := 64-r	// left shift
+
+	z[0] = (x[0] >> r) | (x[1] << l)
+	z[1] = (x[1] >> r) | (x[2] << l)
+	z[2] = (x[2] >> r) | (x[3] << l)
+	z[3] = (x[3] >> r) | (x[4] << l)
+	z[4] = (x[4] >> r)
+
+	return
+}
+
+// reciprocal computes a 320-bit value representing 1/m
+//
+// Notes:
+// - reciprocal(0) = 0
+// - reciprocal(1) = 2^320-1
+// - otherwise, the result is normalised to have non-zero most significant word
+// - starts with a 32-bit division, refines with newton-raphson iterations
+
+func reciprocal(m Int) (mu [5]uint64) {
```

Comment:
As this implementation is .. fairly nontrivial ... it would be good if you can add some links to the origin of the algorithm, so the astute reader can have something to verify against

Reply:
I created this algorithm for this specific purpose, so there is no paper or website to refer to as the origin of the algorithm.
It's really only computing the reciprocal of the modulus with 320-bit precision, taking care to avoid overflow.